lib/exclusive-lock.js

```diff
@@ -0,0 +1,157 @@
+'use strict'
+
+const Ajv = require('ajv')
+const schemas = require('./schemas/index.js')
+const jsonParseFilter = require('./json-parse-filter.js')
+
+const DEFAULT_LOCK_TTL_MS = 3000
+const DEFAULT_LOCK_REFRESH_MS = 1000
+const REQUIRED_MIN_MS_DIFF = 500
+
+const ajv = new Ajv()
+const validate = ajv.compile(schemas.instantiation_options)
+
+class ExclusiveLock {
+  #lock_name
+
+  constructor(opts) {
+    opts = opts || {}
+
+    if (!opts.log?.error) {
+      const err = new Error('A pino logger instance is required')
```

Comment:
eh. the presence of an `error` property doesn't make it a pino instance. 
This check should be better if it is going to be here at all

Reply:
This one is tough. It doesn't have to be pino, but it needs to support the log methods that are used in the code. I got lazy on this. There's a more explicit way I can test for what is required.

Reply:
Use `abstract-logging` -- https://npm.im/abstract-logging:

```js
if (!options.logger) {
  logger = require('abstract-logging')
  loger.child = () => logger
}
```

Otherwise, you get to implement https://github.com/fastify/fastify/blob/bd1f99cb09409d9566aa5e6a9d670f592801f655/lib/logger.js#L103-L117

Reply:
Heh, I did already implement that, actually!  Not fun.  Great suggestion here to fix it with `abstract-logging`

Reply:
I guess it's not technically "fixed". You just get to assume a logger exists that conforms to the interface and write "must be a compatible logger" in the docs, pushing the error responsibility onto the user.

Reply:
Yep, it's better this way.  I changed it, thanks!